### PR TITLE
Implement ldap integration

### LINKED ISSUE #6: Use juju secret for password config
We should start using user defined juju secrets for the bind password

### PR DESCRIPTION
**Pull request by nsklikas**
IAM-999

Closes #6

### CODE REVIEW

File: tests/integration/test_charm.py [hunk 1]
```diff
@@ -3,32 +3,78 @@
...
+    ops_test: OpsTest, ldap_integrator_application: Application, ldap_integrator_charm_config: Dict
+) -> None:
+    await ldap_integrator_application.set_config(ldap_integrator_charm_config)
+    await ops_test.model.integrate(GLAUTH_APP, APP_NAME)
+
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME, GLAUTH_APP], status="active", raise_on_blocked=False, timeout=1000
+    )
+
+    data = await get_app_integration_data(ops_test, GLAUTH_APP, "ldap-client")
```
Comment (wood-push-melon) [inline tests/integration/test_charm.py:64]: Just a question: `ldap-client` is the integration name at the `glauth-k8s` end, right?

File: src/utils.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    return wrapped
+
+
+ldap_integration_exists = integration_existence(LDAP_INTEGRATION_NAME)
+
+
+def missing_config(charm: "LdapIntegratorCharm") -> Set[str]:
+    """Check whether the required configuration has been provided."""
+    required_keys = ["urls", "base_dn", "starttls", "bind_dn", "bind_password"]
+    return {k for k in required_keys if charm.config.get(k, None) is None}
```
Comment (wood-push-melon) [inline src/utils.py:34]: Just two minor nitpicks (feel free to ignore):

- `required_keys` could be declared as a contestant with a tuple or set. So that it's easier to be discovered
- Do we allow empty string for the each config? If not, we could `if not charm.config.get(k)`

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -3,32 +3,78 @@
...
+            application_name=APP_NAME,
+            trust=True,
+        ),
+        await ops_test.model.deploy(
+            GLAUTH_APP,
+            application_name=GLAUTH_APP,
+            channel="edge",
+            trust=True,
+        ),
+        await ops_test.model.deploy(
```
Comment (wood-push-melon) [inline tests/integration/test_charm.py:41]: We don't need to `await` these two coroutines.

File: src/charm.py [hunk 4]
```diff
@@ -16,11 +20,45 @@ class LdapIntegratorCharm(ops.CharmBase):
...
+            bind_dn=self.config.get("bind_dn"),
+            bind_password=self.config.get("bind_password"),
+            auth_method=self.config.get("auth_method"),
+        )
+
+        self.ldap.update_relations_app_data(
+            data, relation_id=ldap_integration,
+        )
+
+    def _on_collect_status(self, event: ops.EventBase) -> None:
```
Comment (wood-push-melon) [inline src/charm.py:53]: Nitpick: `CollectStatusEvent`?

File: tests/integration/test_charm.py (hunk 1 above)
Comment (nsklikas) [reply tests/integration/test_charm.py:64]: correct

**Review by shipperizer — APPROVED** [review]

File: src/charm.py [hunk 5]
```diff
@@ -16,11 +20,46 @@ class LdapIntegratorCharm(ops.CharmBase):
...
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
+        """The central management of the charm operator's status."""
+        if not ldap_integration_exists(self):
+            event.add_status(ops.BlockedStatus(f"Missing integration {LDAP_INTEGRATION_NAME}"))
+
+        if missing := missing_config(self):
+            event.add_status(ops.BlockedStatus(f"Missing required config: {(', ').join(missing)}"))
+
+        event.add_status(ops.ActiveStatus())
```
Comment (BarcoMasile) [inline src/charm.py:62]: question: doesn't this get executed always? meaning, doesn't this always end up with active status?

Comment (nsklikas) [reply src/charm.py:62]: See https://ops.readthedocs.io/en/latest/index.html#ops.CollectStatusEvent

TLDR: we add a bunch of statuses and let the ops framework evaluate the final status. The ActiveStatus has the lowest priority, meaning that if any other status was added, it will not be exposed.

**Review by BarcoMasile — APPROVED** [review]

**Comment by nsklikas** [comment]
I changed it so that now we use juju secrets to provide the bind password to the charm and I updated the README. Please have another look

**Review by BarcoMasile — DISMISSED** [review]

**Review by shipperizer — DISMISSED** [review]

**Review by BarcoMasile — APPROVED** [review]

**Review by BarcoMasile — APPROVED** [review]
